core/src/apps/zcash/orchard/signer.py

```diff
@@ -0,0 +1,300 @@
+from .crypto.note import Note
+from .debug import watch_gc_async
+from .keychain import OrchardKeychain
+from .random import BundleShieldingRng
+
+if TYPE_CHECKING:
+    from typing import Awaitable, List
+    from apps.common.coininfo import CoinInfo
+    from apps.bitcoin.sign_tx.tx_info import TxInfo
+    from .crypto.keys import FullViewingKey
+    from ..approver import ZcashApprover
+    from .random import ActionShieldingRng
+
+
+OVERWINTERED = const(0x8000_0000)
+FLAGS = const(0b0000_0011)  # spends enbled and output enabled
+MAX_SILENT_ORCHARD_INPUTS = const(8)
+
+
+def skip_if_empty(func):
+    """
+    A function decorated by this will not be evaluated,
+    if the Orchard bundle is impty.
+    """
+
+    async def wrapper(self):
+        if self.actions_count == 0:
+            return
+        else:
+            await func(self)
+
+    return wrapper
+
+
+class OrchardSigner:
+    def __init__(
+        self,
+        tx_info: TxInfo,
+        seed: bytes,
+        approver: ZcashApprover,
+        coin: CoinInfo,
+        tx_req: TxRequest,
+    ) -> None:
+        assert tx_req.serialized is not None  # typing
+
+        self.inputs_count = tx_info.tx.orchard_inputs_count
+        self.outputs_count = tx_info.tx.orchard_outputs_count
+
+        if self.inputs_count + self.outputs_count > 0:
+            self.actions_count = max(
+                2,  # minimal required amount of actions
+                self.inputs_count,
+                self.outputs_count,
+            )
+        else:
+            self.actions_count = 0
+
+        if self.actions_count == 0:
+            return  # no need to initialize other attributes
+
+        self.tx_info = tx_info
+        self.keychain = OrchardKeychain.from_seed_and_coin(seed, coin)
+        self.approver = approver
+        self.coin = coin
+        self.tx_req = tx_req
+        assert isinstance(tx_info.sig_hasher, ZcashHasher)
+        self.sig_hasher: ZcashHasher = tx_info.sig_hasher
+
+        account = tx_info.tx.account
+        assert account is not None  # typing
+        key_path = [
+            32 | HARDENED,  # ZIP-32 constant
+            coin.slip44 | HARDENED,  # purpose
+            account | HARDENED,  # account
+        ]
+        self.key_node = self.keychain.derive(key_path)
+
+        self.msg_acc = MessageAccumulator(
+            self.keychain.derive_slip21(
+                [b"Zcash Orchard", b"Message Accumulator"],
+            ).key()
+        )
+
+        self.rng = None
+
+    @skip_if_empty
+    async def process_inputs(self) -> None:
+        await self.check_orchard_inputs_count()
+        for i in range(self.inputs_count):
+            txi = await self.get_input(i)
+            self.msg_acc.xor_message(txi, i)  # add message to the accumulator
+            self.approver.add_orchard_input(txi)
+
+    def check_orchard_inputs_count(self) -> Awaitable[None]:  # type: ignore [awaitable-is-generator]
+        if self.inputs_count > MAX_SILENT_ORCHARD_INPUTS:
```

Comment:
so , there exist a spend linkability _dust attack_ described in §6 of

[Privacy Aspects and Subliminal Channels in Zcash](https://orbilu.uni.lu/bitstream/10993/41278/1/Post_sapling_ZC_paper.pdf)

tldr : attacker sends funds using huge amount of utxos . since number of shielded inputs spent within a transaction is public , attacker can detect that funds were spent by observing a transaction with unusual count of inputs

number 10 is taken from the paper

warning instead of aborting , because this attack is not severe

Reply:
I should document this somewhere

Reply:
e5d373afd2201a3c793ad4288554f0ac3a1763d3